## 1. Fundamental
Don’t get confused by its name! It is **a classification not a regression** algorithm. It is used to estimate discrete values ( Binary values like 0/1, yes/no, true/false ) based on given set of independent variable(s). In simple words, it **predicts the probability of occurrence of an event** by fitting data to a logit function. Hence, it is also known as logit regression. Since, it predicts the probability, its output values lies between 0 and 1 (as expected).

Again, let us try and understand this through a simple example.

*Let’s say your friend gives you a puzzle to solve. There are only 2 outcome scenarios – either you solve it or you don’t. Now imagine, that you are being given wide range of puzzles/quizzes in an attempt to understand which subjects you are good at. The outcome to this study would be something like this – if you are given a trignometry based tenth grade problem, you are 70% likely to solve it. On the other hand, if it is grade fifth history question, the probability of getting an answer is only 30%. This is what Logistic Regression provides you.*

Coming to the math, the log odds of the outcome is modeled as a linear combination of the predictor variables.

$$
\text{odds} = \frac{p}{1-p} = \frac{\text{probability of event occurrence}}{\text{probability of not event occurrence}}\\
\ln(\text{odds})=\ln\left(\frac{p}{1-p}\right)\\
\text{logit}(p)=\ln\left(\frac{p}{1-p}\right)=b_0+b_1x_1+b_2x_2+b_3x_3+....+b_kx_k
$$

Above, $p$ is the **probability of presence** of the characteristic of interest. It chooses parameters that **maximize the likelihood** of observing the sample values rather than that minimize the sum of squared errors (like in ordinary regression).

Now, you may ask, why take a log? For the sake of simplicity, let’s just say that this is one of the best mathematical way to replicate a step function. I can go in more details, but that will beat the purpose of this article.
![](https://www.analyticsvidhya.com/wp-content/uploads/2015/08/Logistic_Regression.png)

**Below are sample code with breast cancer data set**

## 2. Example

In [1]:
# import the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# split tool
from sklearn.model_selection import train_test_split
# Logistic model
from sklearn.linear_model import LogisticRegression
# Evaluation tool
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve

### 2.1 Load the dataset

In [2]:

from sklearn.datasets import load_breast_cancer
breast_cancer = load_breast_cancer()
type(breast_cancer)

sklearn.utils.Bunch

In [3]:
# see what inside
breast_cancer.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

In [4]:
# the dataset description
print(breast_cancer.DESCR)

Breast Cancer Wisconsin (Diagnostic) Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 3 is Mean Radius, field
        13 is Radius SE, field 23 is Worst Radius.

        

There are two values in target data, where 0 represent WDBC-Malignant class and 1 represent WDBC-Benign

In [5]:
# dataframe view
pd.options.display.max_columns = 99
breast_cancer_df = pd.DataFrame(
    data=np.append(breast_cancer.data, breast_cancer.target.reshape(-1, 1), axis=1), 
    columns=np.append(breast_cancer.feature_names,['class'])
)
# Map class value to name for comfortable view
breast_cancer_df['class'] = breast_cancer_df['class']\
                             .map({0: breast_cancer.target_names[0], 1: breast_cancer.target_names[1]})
breast_cancer_df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0         17.990         10.38          122.80     1001.0          0.11840   
1         20.570         17.77          132.90     1326.0          0.08474   
2         19.690         21.25          130.00     1203.0          0.10960   
3         11.420         20.38           77.58      386.1          0.14250   
4         20.290         14.34          135.10     1297.0          0.10030   
5         12.450         15.70           82.57      477.1          0.12780   
6         18.250         19.98          119.60     1040.0          0.09463   
7         13.710         20.83           90.20      577.9          0.11890   
8         13.000         21.82           87.50      519.8          0.12730   
9         12.460         24.04           83.97      475.9          0.11860   
10        16.020         23.24          102.70      797.8          0.08206   
11        15.780         17.89          103.60      781.0          0.09710   
12        19.170         24.80          132.40     1123.0          0.09740   
13        15.850         23.95          103.70      782.7          0.08401   
14        13.730         22.61           93.60      578.3          0.11310   
15        14.540         27.54           96.73      658.8          0.11390   
16        14.680         20.13           94.74      684.5          0.09867   
17        16.130         20.68          108.10      798.8          0.11700   
18        19.810         22.15          130.00     1260.0          0.09831   
19        13.540         14.36           87.46      566.3          0.09779   
20        13.080         15.71           85.63      520.0          0.10750   
21         9.504         12.44           60.34      273.9          0.10240   
22        15.340         14.26          102.50      704.4          0.10730   
23        21.160         23.04          137.20     1404.0          0.09428   
24        16.650         21.38          110.00      904.6          0.11210   
25        17.140         16.40          116.00      912.7          0.11860   
26        14.580         21.53           97.41      644.8          0.10540   
27        18.610         20.25          122.10     1094.0          0.09440   
28        15.300         25.27          102.40      732.4          0.10820   
29        17.570         15.05          115.00      955.1          0.09847   
..           ...           ...             ...        ...              ...   
539        7.691         25.44           48.34      170.4          0.08668   
540       11.540         14.44           74.65      402.9          0.09984   
541       14.470         24.99           95.81      656.4          0.08837   
542       14.740         25.42           94.70      668.6          0.08275   
543       13.210         28.06           84.88      538.4          0.08671   
544       13.870         20.70           89.77      584.8          0.09578   
545       13.620         23.23           87.19      573.2          0.09246   
546       10.320         16.35           65.31      324.9          0.09434   
547       10.260         16.58           65.85      320.8          0.08877   
548        9.683         19.34           61.05      285.7          0.08491   
549       10.820         24.21           68.89      361.6          0.08192   
550       10.860         21.48           68.51      360.5          0.07431   
551       11.130         22.44           71.49      378.4          0.09566   
552       12.770         29.43           81.35      507.9          0.08276   
553        9.333         21.94           59.01      264.0          0.09240   
554       12.880         28.92           82.50      514.3          0.08123   
555       10.290         27.61           65.67      321.4          0.09030   
556       10.160         19.59           64.73      311.7          0.10030   
557        9.423         27.88           59.26      271.3          0.08123   
558       14.590         22

### 2.2 Build the model

In [6]:
# split train/test
Xtrain, Xtest, Ytrain, Ytest = \
train_test_split(
    breast_cancer.data, breast_cancer.target,
    test_size = 0.3, 
    random_state=42
)

In [7]:
# shape of train and test sets
print(Xtrain.shape, Xtest.shape, Ytrain.shape, Ytest.shape)

(398, 30) (171, 30) (398,) (171,)


In [8]:
# Initialize and train model
clf = LogisticRegression()
clf.fit(X=Xtrain, y=Ytrain)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [9]:
# get prediction
# take first row in feature matrix to predict
sample = Xtrain[0:1,:]
prediction = clf.predict(sample)
print('Predicted class:\t', prediction.squeeze(),'-- class name:', breast_cancer.target_names[prediction.squeeze()])
print('True class:\t\t', Ytrain[0], '-- class name:', breast_cancer.target_names[Ytrain[0].squeeze()])

Predicted class:	 1 -- class name: benign
True class:		 1 -- class name: benign


## 3. Evaluate classification model

### 3.1 Accuracy
Accuracy represents the portion that the model can predict the true label
$$
\text{Accuracy} = \frac{\text{number of right predicitons}}{\text{number of totals predicitons}}
$$

In [10]:
# Caculate accuracy by defination
Ytrain_hat = clf.predict(Xtrain)
train_acc_d = np.sum(Ytrain_hat == Ytrain)/float(len(Ytrain))
print("Accuracy on train set:", round(train_acc_d*100,1), '%')
Ytest_hat = clf.predict(Xtest)
test_acc_d = np.sum(Ytest_hat == Ytest)/float(len(Ytest))
print("Accuracy on test set:", round(test_acc_d*100,1), '%')

Accuracy on train set: 95.2 %
Accuracy on test set: 96.5 %


In [11]:
# Calculate accuracy by built-in function
train_acc = lrClass.score(X=Xtrain, y=Ytrain)
test_acc = lrClass.score(X=Xtest, y=Ytest)
print("Accuracy on train set:", round(train_acc*100,1), '%')
print("Accuracy on test set:", round(test_acc*100,1), '%')

NameError: name 'lrClass' is not defined

### 3.2 Confusion Matrix
[wiki link](https://en.wikipedia.org/wiki/Confusion_matrix)

A confusion matrix is a table that is often used to describe the performance of a classification model (or “classifier”) on a set of test data for which the true values are known. 

In [ ]:
# get predictions 
Ytrain_pred = lrClass.predict(Xtrain) # on train set
Ytest_pred = lrClass.predict(Xtest)   # on test set

In [ ]:
# calculate confusion matrices
cm_train = confusion_matrix(y_true=Ytrain, y_pred=Ytrain_pred) # on train set
cm_test = confusion_matrix(y_true=Ytest, y_pred=Ytest_pred)    # on test set 

In [ ]:
# visualization with seaborn
plt.figure(figsize=(10,5))
# on train set
plt.subplot(1,2,1)
sns.heatmap(cm_train, annot=True, fmt='.0f', linewidths=.5, square = True, cmap = 'Blues_r', cbar=False);
plt.ylabel('Actual label');
plt.xlabel('Predicted label');
all_sample_title = 'Accuracy Score on train: %.3f'%(train_acc)
plt.title(all_sample_title, size = 12);

# on test set
plt.subplot(1,2,2)
sns.heatmap(cm_test, annot=True, fmt='.0f', linewidths=.5, square = True, cbar=False);
plt.ylabel('Actual label');
plt.xlabel('Predicted label');
all_sample_title = 'Accuracy Score on test: %.3f' % test_acc
plt.title(all_sample_title, size = 12);

### 3.3 Precision, recall, F-measure and support
![](https://cdn-images-1.medium.com/max/1200/1*g5zpskPaxO8uSl0OWT4NTQ.png)

The **precision** is the ratio $\frac{tp}{tp + fp}$ where $tp$ is the number of true positives and $fp$ the number of false positives. *The precision is intuitively the ability of the classifier to* ***not label a sample as positive if it is negative***.

The **recall** is the ratio $\frac{tp}{tp + fn}$ where $tp$ is the number of true positives and $fn$ the number of false negatives. *The recall is intuitively the ability of the classifier to* ***find all the positive samples***.

The **F-beta score** can be interpreted as a **weighted harmonic mean of the precision and recall**, where an F-beta score reaches its best value at 1 and worst score at 0.

The F-beta score ($F_\beta$) weights the recall more than the precision by a factor of $\beta$, $\beta$ = 1.0 means recall and precision are equally important.

The **support** is the number of occurrences of each class in true label set.

In [ ]:
print('='*60, '\nOn train set:\n')
print(classification_report(Ytrain, Ytrain_pred))
print('='*60, '\nOn test set:\n')
print(classification_report(Ytest, Ytest_pred))

### 3.4 ROC Curve
The [receiver operating characteristic (ROC) curve](https://en.wikipedia.org/wiki/Receiver_operating_characteristic) is another common tool used with binary classifiers. The dotted line represents the ROC curve of a purely random classifier; a good classifier stays as far away from that line as possible (toward the top-left corner).

In [ ]:
plt.figure(figsize=(12,5))
# train set 
plt.subplot(1,2,1)
logit_roc_auc = roc_auc_score(Ytrain, lrClass.predict(Xtrain))
fpr, tpr, thresholds = roc_curve(Ytrain, lrClass.predict_proba(Xtrain)[:,1])
plt.plot(fpr, tpr, label='Logistic Regression (area = %0.2f)' % logit_roc_auc)
plt.plot([0, 1], [0, 1],'r--')
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel('False Positive Rate')  # = FP/(FP+TN)
plt.ylabel('True Positive Rate')   # = recall, sensitive = TP/(TP+FN)
plt.title('Receiver operating characteristic on train')
plt.legend(loc="lower right")

# test set
plt.subplot(1,2,2)
logit_roc_auc = roc_auc_score(Ytest, lrClass.predict(Xtest))
fpr, tpr, thresholds = roc_curve(Ytest, lrClass.predict_proba(Xtest)[:,1])
plt.plot(fpr, tpr, label='Logistic Regression (area = %0.2f)' % logit_roc_auc)
plt.plot([0, 1], [0, 1],'r--')
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('Receiver operating characteristic on test')
plt.legend(loc="lower right")

plt.show()

## 4. Conclusion
In breast cancer prediction, we can build a good classification model by logistic regression that can reach 96% accuracy on test set.